packages/db/postgres/functions.sql

```diff
@@ -293,3 +294,32 @@ FROM public.aggregate_entry ae
 WHERE ae.cid_v1 = ANY (cids)
 ORDER BY de.entry_last_updated
 $$;
+
+CREATE OR REPLACE FUNCTION publish_name_record(data json) RETURNS TEXT
+    LANGUAGE plpgsql
+    volatile
+    PARALLEL UNSAFE
+AS
+$$
+BEGIN
+  INSERT INTO name (key, record, has_v2_sig, seqno, validity)
+  VALUES (data ->> 'key',
+          data ->> 'record',
+          (data ->> 'has_v2_sig')::BOOLEAN,
+          (data ->> 'seqno')::BIGINT,
+          (data ->> 'validity')::BIGINT)
+  ON CONFLICT (key) DO UPDATE
+    SET record = data ->> 'record',
+        has_v2_sig = (data ->> 'has_v2_sig')::BOOLEAN,
+        seqno = (data ->> 'seqno')::BIGINT,
+        validity = (data ->> 'validity')::BIGINT,
+        updated_at = TIMEZONE('utc'::TEXT, NOW())
+    WHERE
+        -- https://github.com/ipfs/go-ipns/blob/a8379aa25ef287ffab7c5b89bfaad622da7e976d/ipns.go#L325
+        ((data ->> 'has_v2_sig')::BOOLEAN = TRUE AND name.has_v2_sig = FALSE) OR
+        ((data ->> 'has_v2_sig')::BOOLEAN = name.has_v2_sig AND (data ->> 'seqno')::BIGINT > name.seqno) OR
+        ((data ->> 'has_v2_sig')::BOOLEAN = name.has_v2_sig AND (data ->> 'seqno')::BIGINT = name.seqno AND (data ->> 'validity')::BIGINT > name.validity) OR
```

Comment:
Not sure if I follow, in https://github.com/ipfs/go-ipns/blob/a8379aa25ef287ffab7c5b89bfaad622da7e976d/ipns.go#L325 a new record will be used if the seqno's are both the same and the sigs are both the same but the `validity` of the new record is greater than the existing one.

Reply:
That is odd. No new records should be created with the same seq number as far as I understand.

I would like to know why this is a possibility, but I guess we can follow what go does for now

Reply:
You implemented the same here: https://github.com/ipfs/js-ipns/blob/master/src/index.js#L481-L503

Reply:
If we're accepting IPNS records, then our rules to update our cache should be the same as in IPFS because:

1. There's not anything stopping people from crafting their own IPNS records and using validity time not sequence to manage updates. So our server code needs to be prepared for that...or we need to lock the server down further to only accept records that increment the sequence number but...
2. My prediction is that these records might not always come through the Web3.Storage API. If we start accepting record updates over pubsub, then I want the behaviour to be the same as in IPFS.

I want Web3.Storage to have the same behaviour as IPFS. I don't want to find a bug in the far future where we're not updating our cache because we decided not to do the same thing as IPFS. One of the biggest parts of this is that it's compatible with IPFS, when we deviate from that it starts to look like a second class citizen. I'm not trying to build a better IPNS, I'm just trying to make a bridge to what's already there that people can use.

Reply:
Those are fare points! Yet, I would still invite designing practical and intuitive system that can help us improve impractical one instead of constraining design by it.

> I consider IPNS impractical because as far as I know no one has adopted it due to it's impracticality (or at least that is my impression).

It could be that there are good reasons for IPNS to do things this way & it's impracticality is caused by other issues. If so we should probably better reflect that in the client as well. In my experience the worst experience is when API imposes set of invariants, but then you discover that somehow those invarians weren't upheld (reminds me of fauna unique index).